python-package/xgboost/sklearn.py

```diff
@@ -801,8 +816,19 @@ def fit(self, X, y, sample_weight=None, base_margin=None,
             else:
                 xgb_options.update({"eval_metric": eval_metric})
 
-        self._le = XGBoostLabelEncoder().fit(y)
-        training_labels = self._le.transform(y)
+        if self.use_label_encoder and not (_is_cudf_df(y) or _is_cudf_ser(y) or _is_cupy_array(y)):
```

Comment:
Why is `and not (_is_cudf_df(y) or _is_cudf_ser(y) or _is_cupy_array(y))` necessary?

Reply:
It's because we cannot use the label encoder at all when the input is cuDF or cuPy.

Reply:
Would it be better if we just deprecate it regardless the input:

```
if self.user_label_encoder:
    warning.warn(....)
```

Reply:
For cuDF input, there is nothing to "deprecate," since previous `XGBClassifier` was not working with a cuDF input. Or do you mean that we should always display the warning?

Reply:
We still have to check the type of the label because inserting cuDF into the label encoder will result in an error.

Reply:
I think there are a few things we need to do.

- Document the pipeline on classification.
- Always give a warning when `use_label_encoder` is set to true.
- Add test cases when the input does not conform to our defined pipeline. (error tests)

Reply:
How should we handle `use_label_encoder=True` with cuDF input? Right now, `XGBClassifier` will silently drop the label encoder to accommodate the cuDF input. Should it be an explicit error if the user sets `use_label_encoder=True` with a cuDF input? (Note that `use_label_encoder=True` is the default setting, to avoid breaking existing code.)

Reply:
> Add test cases when the input does not conform to our defined pipeline

We can add an explicit check as to whether the list of classes (`np.unique(y)`) is actually 0, 1, ..., `[num_class]-1`, and throw an error otherwise.

Reply:
> Should it be an explicit error if the user sets use_label_encoder=True with a cuDF input?

Yes.

> We can add an explicit check as to whether the list of classes (np.unique(y)) is actually 0, 1, ..., [num_class]-1, and throw an error otherwise.

Sounds good.

Reply:
Eh... If we throw errors for that, then the default setting (use_label_encoder=True) will result in an error for all cuDF inputs. I'd prefer keeping the PR as it currently stands, so that unsuspecting users can use fit() with cuDF inputs.

Reply:
User of XGBoost using `pandas` before but switch to cuDF might find the behavior surprising as they are not consistent.

Reply:
Got it. Let's throw an error when `use_label_encoder=True` is set and the input is cuDF.